README.md
# Aera
A Super-Simple HTTP Server library that makes use of functional programming paradigms, but stays unopinionated about how you handle your requests and responses.
___

[![Build Status](https://travis-ci.org/danielkov/aera.svg?branch=master)](https://travis-ci.org/danielkov/aera) [![Coverage Status](https://coveralls.io/repos/github/danielkov/aera/badge.svg?branch=development)](https://coveralls.io/github/danielkov/aera?branch=development) [![Dependencies](https://david-dm.org/danielkov/aera.svg)](https://david-dm.org/danielkov/aera) [![devDependencies Status](https://david-dm.org/danielkov/aera/dev-status.svg)](https://david-dm.org/danielkov/aera?type=dev) [![npm](https://img.shields.io/npm/v/aera.svg?style=flat-square)](https://npmjs.com/package/aera)

```js
const Aera = require('aera')

const app = new Aera()
app.get('/', () => 'Hello, World!')
```

See [the change log](changelog.md) for changes in this version!

## Basic usage

Install and add to your dependencies with the following line:

```sh
npm install --save aera
```

Below you can see a `Hello, World!` example, which is possibly the tiniest web application you can create.

```js
const Aera = require('aera')
const server = new Aera() // Default port is 3000

server.get('/', () => 'Hello, World!')
```

In Aera, you can pass a single function to a path handler and the return value of that function will be reflected in the response. Aera will attempt to make a few guesses, based on the content you provide, for example, if you return an object, it will be stringified and sent with the Content-Type: `application/json`.

Here are the types of return values you can pass to Aera at the moment:

  - String
  - Number
  - Promise
  - Object
  - Readable Stream
  - ... more to come

You can implement a static file server in a single line!

```js
server.get('/:file', ({ params }) => fs.createReadStream(`./my/folder/${params.file}`))
```

You can execute any function that returns any of the supported values and return the value of that, for example: template functions.

```js
server.get('/template', () => myTemplateFunction({name: 'Daniel', status: 'Awesome'}))
```

CRUD in a single line...

```js
server.get('/api/pets/:name', ({ params }) => db.find({ name: params.name })) // given that your db implementation returns a promise.
```

You can also play around with the data and then use it:

```js
server.get('/users/:id', ({ params }) => db.find({ _id: params.id }).then(formatUser).then(renderTemplate))
```

Simple request body echo server:

```js
server.post('/echo', (req) => req)
```

Creating Web Apps should be simple and fun.

## Exceptions and error handling

Aera tries to handle your exceptions for you, but it is completely fine if you want to have your own custom messages for exceptions. You can specify them in the options parameter of the `Aera()` constructor.

```js
const server = new Aera({port: 3000, notFoundException: 'Not the droids you are looking for. Sorry.'})
```

Of course if you wanted to make this nice, you can pass it a template function instead. That would also as long as it returns a valid string value.

```js
const server = new Aera({port: 3000, notFoundException: myNotFoundTemplate()})
```

Currently the following errors will be returned by Aera by default:

  - `notFoundException` status: 404, body: `Not found.` - returned when no path matches the request url.
  - `fileNotFoundException` status: 404, body: `File not found.` - returned when a stream is passed in and it errors.
  - `internalServerErrorException` status: 500, body: `Internal server error.` - returned when the handler has an uncaught exception.
  - `methodNotAllowedException` status: 405, body: `Method not allowed.` - returned when no handlers are available for the method of the path.

## Request and Response

In Aera, the `request` and `response` parameters are actually the native Node JS HTTP request and response objects. If you know how to use native HTTP, you'll know how to use Aera. Read up on the HTTP docs, [here](https://nodejs.org/api/http.html), or look for a guide in the guides folder.

## Running tests

You can run tests with the following line:

```sh
npm test
```

A little advice: in most use cases you can write tests for your application logic, without even involving Aera. This is because Aera encourages the concept of simple functional programming over side-effect-based middleware writing. Create your handlers, run tests on them and then plug them into Aera to see your content appear on your website or service.

## Future

The API is not yet evolved fully, however in the not so far future I'm looking into making it more friendly and easier in production use-cases as well.

## Contribution

I welcome contribution to this project in any form. If you'd like to work with me to improve anything (performance, features), feel free to submit an issue and create a PR.

Currently I am looking into ways to efficiently create hooks for certain events that happen, like `pre-routing`, `post-handler`, etc... This may be useful for authentication, normalization, header checks and so on. This will probably get released in the next major version `1.0.0`.


changelog.md
## Version 1.1.4

  1. Implemented same version of `HttpError` as used in [`aera-tools`](https://www.npmjs.com/package/aera-tools).

## Version 1.1.3

  1. Changed error handling logic, so now error messages of all types will go through the same process as other response types, e.g.: passing in a stream as an error message will pipe that stream to the response object.

## Version 1.1.2

  1. Changed the routing logic a little bit. This means that now you have access to `request.query`, which is an object and search, a.k.a.: query parameters will no longer stop a route from working. This is more like a bugfix, rather than a release change, hence the minor version bump.
```js
.get('/', ({query}) => query) // localhost:3000/?something=value will respond with { "something": "value" }
```
  2. Added some more tests for serving files.

## Version 1.1.1 Release notes

Tiny update. Removed a `console.log()` that remained from development.

Implemented Travic CI for deployment and Coveralls for coverage reporting.

## Version 1.1.0 Release notes

With this version the instance methods have been reworked. New ones added: `head()`, `options()`, `patch()` to compliment the full arsenal of HTTP methods. HEAD requests will also be met with the headers of `GET` request should no `HEAD` request handler be provided.

Summary of changes:

  1. Added `head()`
  2. Added `patch()`
  3. Added `options()`
  4. Changed all methods to now also accept and Array as the first argument, which means:

```js
.get(['/', () => 'Hello']) // Is equal in functionality to
.get('/', () => 'Hello')
```

  5. Released [`aera-tools`](https://npmjs.com/package/aera-tools) to compliment Aera's API.

## Version 1.0.0 Release Notes

I am happy to announce the release of the **first major version**. This is the first major version, because I feel like Aera reached a point at which I would gladly recommend it to anyone wanting to get started with Node JS HTTP. Frankly, not a lot has changed since the previous version, however the API is now a lot more consistent with my expectations. For example, if you return errors, they will be handled properly (e.g.: their message will be displayed and their status, if they have any). The new version now also supports Duplex Streams, along with Readable Streams, which means you can build some very concise and insanely fast web applications, taking advantage of Transform Streams. To sum up the notes, here's a little summary:

  1. Fixed a tiny typo in the README

  2. Aera will now attempt to guess the mime type of your streams and send them over via the headers properly.

  3. The source files in the `/lib` folder are now more modular as I've split them up. This makes them easier to debug.

  4. More tests have been written to ensure Aera works as expected.

  5. Promises returned from handler functions will now get the same treatment as regular responses once they've been resolved (parsing and content type detection). This means you can now return a regular object as the return value of a Promise you passed into the handler function, like so:

```js
.get('/', () => new Promise((resolve, reject) => resolve({message: 'Hello'})))
```

**More info:** performance test are coming in the near future, but rest assured they will match all expectations. I have tested some basic examples with some of the more popular frameworks and with native Node JS and so far the results are awesome.

I will try not to bloat the API, like other frameworks do, instead I'm planning on releasing a bunch of tools you can use with Aera to make your web developing experience even more fun. These will come in the near future as `aera-tools` and will include stuff like easy authentication, database integration and more. Using these will stay completely optional, of course.

More documentation and guides are also coming your way!


lib/index.js
const http = require('http')
const ptr = require('path-to-regexp')

const isReadableStream = require('./utils/isReadableStream')
const isDuplexStream = require('./utils/isDuplexStream')
const isPromise = require('./utils/isPromise')

const HttpError = require('./utils/HttpError')
const parseResult = require('./utils/parseResult')

module.exports = class Server {

  /**
   * constructor - for Server class
   *
   * @param  {object|number} options options to the constructor
   * @return {object}         instance of the class
   */
  constructor (options) {
    let opts = options || {}
    if (typeof opts === 'number') {
      this.port = opts
    }
    else if (typeof opts === 'object') {
      this.port = opts.port || 3000
    }
    else {
      throw new Error('Options argument is not of types number or object.')
    }
    this.routes = []
    this.server = http.createServer(this._handler.bind(this))
    .listen(this.port)

    this.exceptions = {
      fileNotFound: opts.fileNotFoundException || 'File not found.',
      notFound: opts.notFoundException || 'Not found.',
      internalServerError: opts.internalServerErrorException || 'Internal server error.',
      methodNotAllowed: opts.methodNotAllowedException || 'Method not allowed.'
    }
  }


  /**
   * _createRoute - creates a route from the given handler function
   *
   * @param  {string} path    path to the route
   * @param  {function} handler function that should be added as a handler
   * @return {undefined}         returns undefined
   */
  _createRoute (path, method, handler) {
    let added = false
    let { routes } = this
    for (let i = routes.length - 1; i >= 0; i--) {
      if (routes[i].path === path) {
        routes[i][method] = handler
        added = true
        break
      }
    }
    if (!added) {
      let newRoute = {
        path,
        keys: []
      }
      newRoute.re = ptr(path, newRoute.keys)
      newRoute[method] = handler
      this.routes.push(newRoute)
    }
  }

  /**
   * _executeRoute - excecutes the route associated with the path
   *
   * @param  {string} path    path to match routes against
   * @param  {string} method  method to check the routes against
   * @param  {object} context the context of the http request event (request and response objects)
   * @return {many}         internal method, returns the return value of the executed path handler or false
   */
  _executeRoute (path, method, context) {
    let noMatch = true
    let { request, response } = context
    let result
    this.routes.forEach(route => {
      let match = path.match(route.re)
      if (match) {
        noMatch = false
        request.params = {}
        if (route[method]) {
          route.keys.forEach((key, i) => {
            request.params[key.name] = match[i + 1]
          })
          result = route[method](request, response)
        }
        else if (method === 'head' && route['get']) {
          route['get'](request, response)
          result = false
        }
        else {
          result = new HttpError(this.exceptions.methodNotAllowed, 405)
        }
      }
    })
    if (noMatch) {
      result = new HttpError(this.exceptions.notFound, 404)
    }
    return result
  }

  /**
   * _createContext - creates a single context object from the request and response
   *
   * @param  {object} request the raw request object of http
   * @param  {object} result  the raw response object of http
   * @return {object}         the combined context object
   */
  _createContext (request, response) {
    return { request, response }
  }

  /**
   * _handler - this is the handler that gets passed into the http server instance
   *
   * @param  {object} req http request
   * @param  {object} res http response
   * @return {undefined}     internal method, returns undefined
   */
  _handler (req, res) {
    let path = req.url
    let method = req.method.toLowerCase()
    let context = this._createContext(req, res)
    let result
    try {
      result = this._executeRoute(path, method, context)
    } catch (e) {
      result = new HttpError(e.message || this.exceptions.internalServerError, e.status || 500)
    }
    let { exceptions } = this
    parseResult({req, res, result, exceptions})
  }

  /**
   * get - adds a path to the route handlers with `GET` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  get (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'get', fn)
    return this
  }

  /**
   * post - adds a path to the route handlers with `POST` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  post (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'post', fn)
    return this
  }

  /**
   * put - adds a path to the route handlers with `PUT` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  put (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'put', fn)
    return this
  }

  /**
   * delete - adds a path to the route handlers with `DELETE` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  delete (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'delete', fn)
    return this
  }

  /**
   * head - adds a path to the route handlers with `HEAD` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  head (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'head', fn)
    return this
  }

  /**
   * options - adds a path to the route handlers with `OPTIONS` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  options (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'options', fn)
    return this
  }

  /**
   * patch - adds a path to the route handlers with `PATCH` method.
   *
   * @param  {string} path string to match the path against
   * @param  {function} fn   handler function to add to the path
   * @return {object}      instance of the class
   */
  patch (...args) {
    let path
    let fn
    if (Array.isArray(args[0])) {
      path = args[0][0]
      fn = args[0][1]
    }
    else {
      path = args[0]
      fn = args[1]
    }
    this._createRoute(path, 'patch', fn)
    return this
  }
}
